pysph/cpy/jit.py

```diff
@@ -0,0 +1,509 @@
+from functools import wraps
+from textwrap import dedent, wrap
+
+from mako.template import Template
+import numpy as np
+import inspect
+import ast
+import importlib
+import warnings
+import sys
+from pytools import memoize
+from .config import get_config
+from .cython_generator import get_parallel_range, CythonGenerator
+from .transpiler import (
+    Transpiler,
+    convert_to_float_if_needed,
+    filter_calls,
+    get_external_symbols_and_calls,
+    BUILTINS)
+from .types import (dtype_to_ctype, get_declare_info,
+                    dtype_to_knowntype, annotate)
+from .extern import Extern
+from .ast_utils import get_unknown_names_and_calls
+
+import pysph.cpy.array as array
+import pysph.cpy.parallel as parallel
+
+
+def kernel_cache_key_args(obj, *args):
+    key = [get_ctype_from_arg(arg) for arg in args]
+    key.append(obj)
+    return tuple(key)
+
+
+def kernel_cache_key_kwargs(obj, **kwargs):
+    key = [get_ctype_from_arg(arg) for arg in kwargs.values()]
+    key.append(obj)
+    return tuple(key)
+
+
+def getargspec(f):
+    getargspec_f = getattr(inspect, 'getfullargspec',
```

Comment:
Instead of doing this each time, you can define this at the module level so the getattr is done once, but this is a minor thing.  Can be done later.